From 6c691e8c481b32afecc16a36500cf8f572d6a3dd Mon Sep 17 00:00:00 2001
From: Megumi_fox <i@megumifox.com>
Date: Sat, 1 Jul 2023 21:42:38 +0800
Subject: [PATCH] Use font from environment variables

Telegram specify opensans as the default font. In some situation, this font make the chat hard read. This commit allow user change the font family and style using environment variables. If variables not set, will use system QT settings default font as fallback.
TELEGRAM_DESKTOP_UI_FONT setting font family
TELEGRAM_DESKTOP_UI_FONT_BOLD_STYLE setting font style to show bold style
TELEGRAM_DESKTOP_UI_FONT_MONO setting mono font family

Signed-off-by: Megumi_fox <i@megumifox.com>
---
 ui/style/style_core_custom_font.cpp | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/ui/style/style_core_custom_font.cpp b/ui/style/style_core_custom_font.cpp
index 8f42baf..302492a 100644
--- a/ui/style/style_core_custom_font.cpp
+++ b/ui/style/style_core_custom_font.cpp
@@ -33,9 +33,25 @@ QFont ResolveFont(const QString &familyOverride, uint32 flags, int size) {
 	const auto italic = (flags & FontItalic);
 	const auto &custom = bold ? BoldFont : RegularFont;
 	const auto useCustom = !custom.family.isEmpty();
+	const QString FontFamilyName = qEnvironmentVariable("TELEGRAM_DESKTOP_UI_FONT");
+	const QString FontFamilyNameMono = qEnvironmentVariable("TELEGRAM_DESKTOP_UI_FONT_MONO");
+	const QString FontFamilyBoldStyle = qEnvironmentVariable("TELEGRAM_DESKTOP_UI_FONT_BOLD_STYLE");
+	const QString FontFamilyNameOverride = FontFamilyName.isEmpty() ? QGuiApplication::font().family() : FontFamilyName;
 
 	auto result = QFont(QGuiApplication::font().family());
-	if (!familyOverride.isEmpty()) {
+	if (!FontFamilyNameOverride.isEmpty()) {
+		result.setFamily(FontFamilyNameOverride);
+		if (bold) {
+			if (FontFamilyBoldStyle.isEmpty()){
+				result.setWeight(QFont::Medium);
+			} else {
+				result.setStyleName(FontFamilyBoldStyle);
+			}
+		}
+		if (flags & FontMonospace) {
+			result.setFamily(FontFamilyNameMono.isEmpty() ? MonospaceFont() : FontFamilyNameMono);
+		}
+	} else if (!familyOverride.isEmpty()) {
 		result.setFamily(familyOverride);
 		if (bold) {
 			result.setBold(true);
-- 
2.41.0

